silx view: Added `--slices` option

t20100 opened the pull request:
This PR adds the option to pass multiple integers for displaying slices with `silx view` through a `--slice` option.

To do so, it adds a `DatasetSlice` class to expose the opened slice as a dataset and makes the necessary changes to make it work.

The content of this PR is adapted from PR #3677 (which was hard to rebase).
It is a simpler version of PR #3677 since it only provides `--slice` and not URL `#fragment` support which could be added later if needed.

t20100 commented on src/silx/gui/hdf5/Hdf5TreeModel.py:
```diff
@@ -61,6 +62,8 @@ def _createRootLabel(h5obj):
         if path.startswith("/"):
             path = path[1:]
         label = "%s::%s" % (filename, path)
+        if isinstance(h5obj, DatasetSlice):
+            label += str(list(h5obj.indices))
```
I'm not a fan of having specific code for `DatasetSlice` here, but I couldn't find a better way, except maybe adding this in the `silx.io.commonh5` API...

t20100 commented on src/silx/gui/hdf5/Hdf5TreeModel.py:
```diff
@@ -573,9 +576,12 @@ def __areH5pyObjectEqual(obj1, obj2):
+            info = [item.name]
+            if item.file is not None:
+                info += [item.file.filename, item.file.mode]
+            if isinstance(item, DatasetSlice):
+                info.append(item.indices)
```
Same, it'd be best to avoid specific code here.
I wanted to rework the way to check equality, but that's complex changes

t20100 commented:
CI fails a lot on Windows with python3.7 and pyside6, with exceptions like:
```
    def tearDown(self):
        self.qapp.processEvents()
    
        if len(self.__class__._exceptions) > 0:
            messages = "\n".join(self.__class__._exceptions)
>           raise AssertionError("Exception occured in Qt thread:\n" + messages)
E           AssertionError: Exception occured in Qt thread:
E           Traceback (most recent call last):
E             File "C:\projects\silx\venv_test\lib\site-packages\silx\gui\data\DataViewer.py", line 290, in __setDataInView
E               if (self.__previousSelection.slice != self.__displayedSelection.slice or
E           AttributeError: 'NoneType' object has no attribute 'slice'
E           Traceback (most recent call last):
E             File "C:\projects\silx\venv_test\lib\site-packages\silx\gui\data\DataViewer.py", line 290, in __setDataInView
E               if (self.__previousSelection.slice != self.__displayedSelection.slice or
E           AttributeError: 'NoneType' object has no attribute 'slice'
E           Traceback (most recent call last):
E             File "C:\projects\silx\venv_test\lib\site-packages\silx\gui\data\DataViewer.py", line 290, in __setDataInView
E               if (self.__previousSelection.slice != self.__displayedSelection.slice or
E           AttributeError: 'NoneType' object has no attribute 'slice'
venv_test\lib\site-packages\silx\gui\utils\testutils.py:188: AssertionError
```

```
    def tearDown(self):
        self.qapp.processEvents()
    
        if len(self.__class__._exceptions) > 0:
            messages = "\n".join(self.__class__._exceptions)
>           raise AssertionError("Exception occured in Qt thread:\n" + messages)
E           AssertionError: Exception occured in Qt thread:
E           Traceback (most recent call last):
E             File "C:\projects\silx\venv_test\lib\site-packages\silx\gui\data\DataViewer.py", line 287, in __setDataInView
E               self.__currentView.setData(self.__displayedData)
E             File "C:\projects\silx\venv_test\lib\site-packages\silx\gui\data\DataViews.py", line 657, in setData
E               self.__updateDisplayedView()
E             File "C:\projects\silx\venv_test\lib\site-packages\silx\gui\data\DataViews.py", line 635, in __updateDisplayedView
E               w = self.__currentView.getWidget()
E             File "C:\projects\silx\venv_test\lib\site-packages\silx\gui\data\DataViews.py", line 351, in getWidget
E               self.__widget = self.createWidget(self.__parent)
E             File "C:\projects\silx\venv_test\lib\site-packages\silx\gui\data\DataViews.py", line 1347, in createWidget
E               widget = RecordTableView(parent)
E           AttributeError: __init__
venv_test\lib\site-packages\silx\gui\utils\testutils.py:188: AssertionError
```

vallsv commented on src/silx/app/view/main.py:
```diff
@@ -44,6 +44,14 @@ def createParser():
+        type=int,
+        nargs='+',
+        help='List of slice indices to open (Only for dataset)',
+    )
```
I think it would be better to use a real slice here

```
--slice 1,2,3
```
instead of
```
--slice 1 2 3
```

As we already have the parser, it's costless

t20100 replied:
`--slice 1 2 3` opens not just one slice (1,2,3) as a dataset but 3 slices.
Maybe best to call it `--slices`.

I haven't done it, but we could also support `--slice 1::4,:10,...,1 3:5,:,1` if needed.... but the current main need is to open slices in a 3D stack.

payno reviewed (approved):
work smoothly on my datasets

t20100 commented:
- Fixed CI: Updated python used to test with pyside6 in order not to restrict matplotlib version.
- Changed option from `--slice` to `--slices`

Finally ready for review/merge

payno reviewed (approved):
still fine with me.
Maybe we could add an example on how to use it on the PR and on the text documentation (`view.rst/ Examples of usage`) as well like:
```
# open specific dataset
silx view 5.06_crayon_W150_60_Al2_W0.25_xc1000_.nx::entry/instrument/detector/data
# open specific set of dataset from name / location
silx view 5.06_crayon_W150_60_Al2_W0.25_xc1000_.nx::entry/*/data
# open specific set of datasets slices (first and last on this example)
silx view 5.06_crayon_W150_60_Al2_W0.25_xc1000_.nx::entry/*/data --slices 0 -1
```
